### PR TITLE
Dev/jacdavis/sequencepoints

### PR DESCRIPTION
**Pull request by jacdavis**

### CODE REVIEW

File: ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs [hunk 1]
```diff
@@ -266,12 +287,12 @@ void EndSequencePoint(TextLocation startLocation, TextLocation endLocation)
...
 				sequencePoints.Add((current.Function, new DebugInfo.SequencePoint {
 					Offset = (int)longSet.Intervals[0].Start,
-					EndOffset = (int)longSet.Intervals[0].End,
+					EndOffset = (int)longSet.Intervals[longSet.Intervals.Count() - 1].End,
 					StartLine = startLocation.Line,
 					StartColumn = startLocation.Column,
 					EndLine = endLocation.Line,
 					EndColumn = endLocation.Column
-				}));
+				})); ;
```
Comment (jacdavis) [inline ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: remove

File: ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs [hunk 2]
```diff
@@ -266,12 +287,12 @@ void EndSequencePoint(TextLocation startLocation, TextLocation endLocation)
 				Debug.Assert(!longSet.IsEmpty);
 				sequencePoints.Add((current.Function, new DebugInfo.SequencePoint {
 					Offset = (int)longSet.Intervals[0].Start,
-					EndOffset = (int)longSet.Intervals[0].End,
+					EndOffset = (int)longSet.Intervals[longSet.Intervals.Count() - 1].End,
```
Comment (jacdavis) [inline ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: I don't recall why this was necessary. Might be able to revert.

File: ICSharpCode.Decompiler/CSharp/ExpressionBuilder.cs [hunk 3]
```diff
@@ -154,7 +154,7 @@ public TranslatedExpression Translate(ILInstruction inst, IType typeHint = null)
 				} else if (inst.ResultType == StackType.Ref) {
 					Debug.Assert(cexpr.Type.GetStackType() == StackType.Ref || cexpr.Type.GetStackType().IsIntegerType());
 				} else {
-					Debug.Assert(cexpr.Type.GetStackType() == inst.ResultType);
+					//Debug.Assert(cexpr.Type.GetStackType() == inst.ResultType);
```
Comment (jacdavis) [inline ICSharpCode.Decompiler/CSharp/ExpressionBuilder.cs]: This was constantly firing while testing with and without my changes. Is it a bogus assert?

File: ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs [hunk 4]
```diff
@@ -297,7 +318,7 @@ void AddToSequencePoint(ILInstruction inst)
 			if (HasUsableILRange(inst) && current.Intervals != null) {
 				current.Intervals.AddRange(inst.ILRanges);
 				var function = inst.Parent.Ancestors.OfType<ILFunction>().FirstOrDefault();
-				Debug.Assert(current.Function == null || current.Function == function);
+				//Debug.Assert(current.Function == null || current.Function == function);
```
Comment (jacdavis) [inline ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: Same as the other assert. This was firing on many binaries without my changes. I disabled it for sanity reasons. Is it bogus?

File: ICSharpCode.Decompiler/IL/Instructions/ILFunction.cs [hunk 5]
```diff
@@ -108,12 +109,17 @@ partial class ILFunction
 		/// </summary>
 		public IMethod MoveNextMethod;
 
+		/// <summary>
+		/// If this function is an iterator/async, this field stores the compiler-generated MoveNext() method.
+		/// </summary>
+		public ILFunction MoveNextFunction;
```
Comment (jacdavis) [inline ICSharpCode.Decompiler/IL/Instructions/ILFunction.cs]: I needed access to this from the sequence point builder to obtain the correct sp candidates

File: global.json [hunk 6]
```diff
@@ -3,6 +3,6 @@
 		"MSBuild.Sdk.Extras": "2.0.54"
 	},
 	"sdk": {
-		"version": "3.1.100"
+		"version": "3.2.000"
```
Comment (jacdavis) [inline global.json]: Revert this before submitting.

File: ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs [hunk 7]
```diff
@@ -348,26 +371,64 @@ internal static bool HasUsableILRange(ILInstruction inst)
...
 					var hidden = new DebugInfo.SequencePoint();
 					hidden.Offset = pos;
 					hidden.EndOffset = function.CodeSize;
 					hidden.SetHidden();
 					newList.Add(hidden);
 				}
+
+				
+				List<int> sequencePointCandidates = function.MoveNextFunction == null ? function.SequencePointCandidates : function.MoveNextFunction.SequencePointCandidates;
+				if (sequencePointCandidates.Count > 2) {
```
Comment (dgrunwald) [inline ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: Why `> 2` and not `> 0` ?
Even a single candidate might help us move one of ours?

File: ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs (hunk 2 above)
Comment (dgrunwald) [inline ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: This might be problematic -- if there's more than one interval, they are not contiguous, so this might put a lot of unrelated code into the sequence point.

Comment (jacdavis) [reply ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: I'm not even sure why I made this change. It was months ago when I first looked at the opening brace issue. I believe I can revert this and will retest afterwards.

File: ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs (hunk 7 above)
Comment (jacdavis) [reply ICSharpCode.Decompiler/CSharp/SequencePointBuilder.cs]: Good catch. It's a copy and replace typo. Should have been if newList.Count > 2

File: ICSharpCode.Decompiler/IL/ILReader.cs [hunk 8]
```diff
@@ -400,6 +405,10 @@ void ReadInstructions(CancellationToken cancellationToken)
 						currentStack = ImmutableStack<ILVariable>.Empty;
 					}
 				}
+
+				if (IsSequencePointInstruction(decodedInstruction)) {
```
Comment (dgrunwald) [inline ICSharpCode.Decompiler/IL/ILReader.cs]: Could use `IsSequencePointInstruction(decodedInstruction) || startedWithEmptyStack` here.
Where you could initialize `bool startedWithEmptyStack = currentStack.IsEmpty;` before the `DecodeInstruction()` call.
